cmd/ctl/pkg/create/certificaterequest/certificaterequest.go

```diff
@@ -0,0 +1,308 @@
+	r := builder.
+		WithScheme(scheme, schema.GroupVersion{Group: cmapiv1alpha2.SchemeGroupVersion.Group, Version: runtime.APIVersionInternal}).
+		LocalParam(true).ContinueOnError().
+		NamespaceParam(o.CmdNamespace).DefaultNamespace().
+		FilenameParam(o.EnforceNamespace, &resource.FilenameOptions{Filenames: []string{o.InputFilename}}).Flatten().Do()
+
+	if err := r.Err(); err != nil {
+		return fmt.Errorf("error when getting Result from Builder: %s", err)
+	}
+
+	singleItemImplied := false
+	infos, err := r.IntoSingleItemImplied(&singleItemImplied).Infos()
+	if err != nil {
+		return fmt.Errorf("error when getting infos out of Result: %s", err)
+	}
+
+	// Ensure only one object per command
+	if len(infos) == 0 {
+		return errors.New("no object passed to create certificaterequest")
+	}
+	if len(infos) > 1 {
+		objects := ""
+		for _, info := range infos {
+			namespace := info.Namespace
+			if namespace == "" {
+				namespace = "default"
+			}
+			objects = objects + fmt.Sprintf("Object with kind %s, name %s, namespace %s\n",
+				info.Object.GetObjectKind().GroupVersionKind().Kind,
+				info.Name,
+				namespace)
+		}
+		return fmt.Errorf("multiple objects passed to create certificaterequest:\n%s", objects)
+	}
+	info := infos[0]
+	// Convert to v1alpha2 because that version is needed for functions that follow
+	crtObj, err := scheme.ConvertToVersion(info.Object, cmapiv1alpha2.SchemeGroupVersion)
+	if err != nil {
+		return fmt.Errorf("failed to convert object into version v1alpha2: %v", err)
+	}
+
+	// Cast Object into Certificate
+	crt, ok := crtObj.(*cmapiv1alpha2.Certificate)
+	if !ok {
+		return errors.New("decoded object is not a v1alpha2 Certificate")
+	}
+
+	signer, err := pki.GeneratePrivateKeyForCertificate(crt)
+	if err != nil {
+		return fmt.Errorf("error when generating new private key for CertificateRequest: %v", err)
+	}
+
+	keyData, err := pki.EncodePrivateKey(signer, crt.Spec.KeyEncoding)
+	if err != nil {
+		return fmt.Errorf("failed to encode new private key for CertificateRequest: %v", err)
+	}
+
+	// Use name for CertificateRequest if specified as arg, else will use name of the Certificate as GenerateName
+	crName := ""
+	if len(args) > 0 {
+		crName = args[0]
+	}
+	req, err := buildCertificateRequest(crt, keyData, crName)
+	if err != nil {
+		return fmt.Errorf("error when building CertificateRequest: %v", err)
+	}
+
+	ns := crt.Namespace
+	if ns == "" {
+		ns = o.CmdNamespace
+	}
+	req, err = o.CMClient.CertmanagerV1alpha2().CertificateRequests(ns).Create(context.TODO(), req, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating CertificateRequest: %v", err)
+	}
+	fmt.Fprintf(o.Out, "CertificateRequest %s has been created in namespace %s\n", req.Name, req.Namespace)
+
+	// Storing private key to file
+	keyFileName := req.Name + ".key"
+	if o.KeyFilename != "" {
+		keyFileName = o.KeyFilename
+	}
+	if err := ioutil.WriteFile(keyFileName, keyData, 0644); err != nil {
```

Comment:
As we were chatting about on the call, `0600` is best for now to protect generated private keys 😄